org.apache.commons.lang3.SerializationException (Waypoint activity)

### Describe your problem!

Google Play report for all versions of current beta phase

41 occurrences, 23 users, Android 8-11



### How to reproduce?

unclear

### Actual result after these steps?

_No response_

### Expected result after these steps?

_No response_

### Reproducible

Unclear

### c:geo Version

2021.07.28-RC

### System information

_No response_

### Additional Information

```
org.apache.commons.lang3.SerializationException: 
  at org.apache.commons.lang3.SerializationUtils.serialize (SerializationUtils.java:140)
  at org.apache.commons.lang3.SerializationUtils.serialize (SerializationUtils.java:154)
  at cgeo.geocaching.ui.dialog.CoordinatesCalculateDialog.onSaveInstanceState (CoordinatesCalculateDialog.java:529)
  at androidx.fragment.app.Fragment.performSaveInstanceState (Fragment.java:3151)
  at androidx.fragment.app.FragmentStateManager.saveBasicState (FragmentStateManager.java:683)
  at androidx.fragment.app.FragmentStateManager.saveState (FragmentStateManager.java:649)
  at androidx.fragment.app.FragmentStore.saveActiveFragments (FragmentStore.java:177)
  at androidx.fragment.app.FragmentManager.saveAllState (FragmentManager.java:2655)
  at androidx.fragment.app.FragmentController.saveAllState (FragmentController.java:152)
  at androidx.fragment.app.FragmentActivity$1.saveState (FragmentActivity.java:133)
  at androidx.savedstate.SavedStateRegistry.performSave (SavedStateRegistry.java:227)
  at androidx.savedstate.SavedStateRegistryController.performSave (SavedStateRegistryController.java:74)
  at androidx.activity.ComponentActivity.onSaveInstanceState (ComponentActivity.java:314)
  at cgeo.geocaching.EditWaypointActivity.onSaveInstanceState (EditWaypointActivity.java:485)
  at cgeo.geocaching.EditWaypointActivity_.onSaveInstanceState (EditWaypointActivity_.java:117)
  at android.app.Activity.performSaveInstanceState (Activity.java:2037)
  at android.app.Instrumentation.callActivityOnSaveInstanceState (Instrumentation.java:1479)
  at android.app.ActivityThread.callActivityOnSaveInstanceState (ActivityThread.java:5305)
  at android.app.ActivityThread.callActivityOnStop (ActivityThread.java:4631)
  at android.app.ActivityThread.handleSleeping (ActivityThread.java:4754)
  at android.app.ActivityThread.access$2500 (ActivityThread.java:220)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1958)
  at android.os.Handler.dispatchMessage (Handler.java:107)
  at android.os.Looper.loop (Looper.java:214)
  at android.app.ActivityThread.main (ActivityThread.java:7397)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:492)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:935)
Caused by: java.io.NotSerializableException: 
  at java.io.ObjectOutputStream.writeObject0 (ObjectOutputStream.java:1240)
  at java.io.ObjectOutputStream.writeObject (ObjectOutputStream.java:354)
  at java.util.ArrayList.writeObject (ArrayList.java:762)
  at java.lang.reflect.Method.invoke (Native Method)
  at java.io.ObjectStreamClass.invokeWriteObject (ObjectStreamClass.java:1036)
  at java.io.ObjectOutputStream.writeSerialData (ObjectOutputStream.java:1552)
  at java.io.ObjectOutputStream.writeOrdinaryObject (ObjectOutputStream.java:1488)
  at java.io.ObjectOutputStream.writeObject0 (ObjectOutputStream.java:1234)
  at java.io.ObjectOutputStream.defaultWriteFields (ObjectOutputStream.java:1604)
  at java.io.ObjectOutputStream.writeSerialData (ObjectOutputStream.java:1565)
  at java.io.ObjectOutputStream.writeOrdinaryObject (ObjectOutputStream.java:1488)
  at java.io.ObjectOutputStream.writeObject0 (ObjectOutputStream.java:1234)
  at java.io.ObjectOutputStream.writeObject (ObjectOutputStream.java:354)
  at org.apache.commons.lang3.SerializationUtils.serialize (SerializationUtils.java:138)
```

--- a/main/src/cgeo/geocaching/calculator/ButtonData.java
+++ b/main/src/cgeo/geocaching/calculator/ButtonData.java
@@ -2,13 +2,15 @@ package cgeo.geocaching.calculator;
 
 import static cgeo.geocaching.models.CalcState.ERROR_CHAR;
 
+import java.io.Serializable;
+
 import org.json.JSONException;
 import org.json.JSONObject;
 
 /**
  * Data used to capture the state of this particular button such that it can be restored again later
  */
-public final class ButtonData implements JSONAble {
+public final class ButtonData implements JSONAble, Serializable {
 
     /** The three states the button can be put into */
     public enum ValueType {

--- a/main/src/cgeo/geocaching/calculator/VariableData.java
+++ b/main/src/cgeo/geocaching/calculator/VariableData.java
@@ -7,6 +7,7 @@ import static cgeo.geocaching.models.CalcState.ERROR_STRING;
 
 import android.content.Context;
 
+import java.io.Serializable;
 import java.util.List;
 import java.util.Locale;
 
@@ -17,7 +18,7 @@ import org.json.JSONObject;
 /**
  * Data used to capture the state of this Variable such that it can be restored again later
  */
-public final class VariableData implements JSONAble {
+public final class VariableData implements JSONAble, Serializable {
     private final char name;
     /**
      * Note, we have to use a String rather than an Editable as Editable's can't be serialized

--- a/tests/src/cgeo/geocaching/calculator/FormulaParserTest.java
+++ b/tests/src/cgeo/geocaching/calculator/FormulaParserTest.java
@@ -6,6 +6,7 @@ import cgeo.geocaching.models.WaypointParser;
 import java.util.ArrayList;
 import java.util.List;
 
+import org.apache.commons.lang3.SerializationUtils;
 import org.junit.Test;
 import static org.assertj.core.api.Assertions.failBecauseExceptionWasNotThrown;
 import static org.assertj.core.api.Java6Assertions.assertThat;
@@ -127,5 +128,17 @@ public class FormulaParserTest {
         } catch (final FormulaParser.ParseException e) {
             // expected
         }
+    }
+
+    @Test
+    public void testSerializeCalcState() {
+        final List<VariableData> variables = new ArrayList<>();
+        variables.add(new VariableData('a', "expression"));
+        final CalcState calcState = CoordinatesCalculateUtils.createCalcState("N  AB° 48.[B+C-A]^2", "E (B%C)°  38.(D+F)*2", variables);
+        calcState.buttons.add(new ButtonData());
+
+        final byte[] serValue = SerializationUtils.serialize(calcState);
+        final CalcState calcState2 = SerializationUtils.deserialize(serValue);
+        assertThat(calcState2.plainLat).isEqualTo(calcState.plainLat);
     }
 }

